### ISSUE TITLE
[BUG] VM with unschedule disks doesn't show clear warning message

### ISSUE THREAD

**Issue by yasker** [harvester/harvester#1695]
**Describe the bug**
VM with unschedule disks doesn't show clear warning message

**To Reproduce**
1. Create a VM that exceed the storage capacity, e.g. 100T
2. VM failed to start, but no error message giving to the user
3. The only available information in the event is

Reason | Resource | Date
-- | -- | --
FailedMount | Pod virt-launcher-test-nzg65Unable to attach or mount volumes: unmounted volumes=[disk-0], unattached volumes=[container-disks sockets disk-0 cloudinitdisk-ndata public ephemeral-disks hotplug-disks libvirt-runtime cloudinitdisk-udata private]: timed out waiting for the condition | 1.3 mins ago
FailedAttachVolume | Pod virt-launcher-test-nzg65AttachVolume.Attach failed for volume "pvc-db775cbb-a5a2-4479-83f6-d7d9af85bfb7" : rpc error: code = Aborted desc = volume pvc-db775cbb-a5a2-4479-83f6-d7d9af85bfb7 is not ready for workloads | 1.3 mins ago
FailedMount | Pod virt-launcher-test-nzg65Unable to attach or mount volumes: unmounted volumes=[disk-0], unattached volumes=[ephemeral-disks private libvirt-runtime cloudinitdisk-udata public container-disks hotplug-disks cloudinitdisk-ndata disk-0 sockets]: timed out waiting for the condition

**Expected behavior**
VM should show similar warning information to the user like `insufficent storage`

**Support bundle**

**Environment:**
 - Harvester ISO version: v1.0.0
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): any

**Additional context**

**Comment by guangbochen** [comment]
There is no obvious message currently showing on the Harvester side due to complicated resource mapping, will make an enhancement on the next version, with the current version you can visit the [Longhorn UI](https://docs.harvesterhci.io/v1.0/troubleshooting/harvester/#access-embedded-longhorn) and check the detailed volume issues accordingly.

**Comment by w13915984028** [comment]
reproduced:
1. volume with massive size
![image](https://user-images.githubusercontent.com/31133476/149165237-c4f91b1f-8fa7-4986-8006-d8c8c0d3454c.png)
2. vm is in "Starting"
![image](https://user-images.githubusercontent.com/31133476/149163669-89767991-d8fa-40a8-8c8a-abeeb58b49b3.png)

3. events reports failure
![image](https://user-images.githubusercontent.com/31133476/149165154-7937f09b-3e6e-4fc5-a895-ff916effbe73.png)




4. embedded` longhorn-ui` doesn´t give enough error message
![image](https://user-images.githubusercontent.com/31133476/149166434-2e227e74-111a-4fb4-984b-8af21cdd4ba7.png)

![image](https://user-images.githubusercontent.com/31133476/149166505-7e754ca4-58b4-43a1-b4ef-bdebf4ea7465.png)


5. it needs knowledge to know from following info that the pvc is in "unknown" state, but is still lack of straightforward info like `"insufficient storage"`



```
harv2:~ # kubectl get volumes.longhorn.io -A
NAMESPACE         NAME                                       STATE      ROBUSTNESS   SCHEDULED   SIZE            NODE    AGE
longhorn-system   pvc-15cc2e31-9ae2-46e8-8d9d-405c77826d53   attached   degraded     False       53687091200     harv2   5d1h
longhorn-system   pvc-2c5f506a-b75c-4028-987b-feaa0a49c58b   attached   degraded     False       10485760        harv2   5d1h

longhorn-system   pvc-3a99612e-7fee-46e2-8b56-df8f88b189ec   detached   unknown      False       1073741824000           9m1s

longhorn-system   pvc-49d2a05a-0c02-4fab-ada3-e3793850bc81   detached   unknown      True        1073741824              17h


harv2:~ # kubectl get volumes.longhorn.io pvc-3a99612e-7fee-46e2-8b56-df8f88b189ec -n longhorn-system
NAME                                       STATE      ROBUSTNESS   SCHEDULED   SIZE            NODE   AGE
pvc-3a99612e-7fee-46e2-8b56-df8f88b189ec   detached   unknown      False       1073741824000          10m
```

harv2:~ #
harv2:~ # kubectl get volumes.longhorn.io pvc-3a99612e-7fee-46e2-8b56-df8f88b189ec -n longhorn-system -o JSON
```
{
    "apiVersion": "longhorn.io/v1beta1",
    "kind": "Volume",
    "metadata": {
        "creationTimestamp": "2022-01-12T14:47:09Z",
        "finalizers": [
            "longhorn.io"
        ],
        "generation": 1,
        "labels": {
            "longhornvolume": "pvc-3a99612e-7fee-46e2-8b56-df8f88b189ec",
            "recurring-job-group.longhorn.io/default": "enabled"
        },
        "name": "pvc-3a99612e-7fee-46e2-8b56-df8f88b189ec",
        "namespace": "longhorn-system",
        "resourceVersion": "4848310",
        "uid": "1066af41-21b3-4605-a8c2-b94966ee71b7"
    },
    "spec": {
        "Standby": false,
        "accessMode": "rwx",
        "backingImage": "default-image-mxcmj",
        "baseImage": "",
        "dataLocality": "disabled",
        "dataSource": "",
        "disableFrontend": false,
        "diskSelector": null,
        "encrypted": false,
        "engineImage": "longhornio/longhorn-engine:v1.2.3",
        "fromBackup": "",
        "frontend": "blockdev",
        "lastAttachedBy": "",
        "migratable": true,
        "migrationNodeID": "",
        "nodeID": "",
        "nodeSelector": null,
        "numberOfReplicas": 3,
        "replicaAutoBalance": "ignored",
        "revisionCounterDisabled": false,
        "size": "1073741824000",
        "staleReplicaTimeout": 30
    },
    "status": {
        "actualSize": 0,
        "cloneStatus": {
            "snapshot": "",
            "sourceVolume": "",
            "state": ""
        },
        "conditions": {
            "restore": {
                "lastProbeTime": "",
                "lastTransitionTime": "2022-01-12T14:47:10Z",
                "message": "",
                "reason": "",
                "status": "False",
                "type": "restore"
            },
            "scheduled": {
                "lastProbeTime": "",
                "lastTransitionTime": "2022-01-12T14:47:10Z",
                "message": "",
                "reason": "ReplicaSchedulingFailure",
                "status": "False",
                "type": "scheduled"
            },
            "toomanysnapshots": {
                "lastProbeTime": "",
                "lastTransitionTime": "2022-01-12T14:47:10Z",
                "message": "",
                "reason": "",
                "status": "False",
                "type": "toomanysnapshots"
            }
        },
        "currentImage": "longhornio/longhorn-engine:v1.2.3",
        "currentNodeID": "",
        "expansionRequired": false,
        "frontendDisabled": false,
        "isStandby": false,
        "kubernetesStatus": {
            "lastPVCRefAt": "",
            "lastPodRefAt": "",
            "namespace": "default",
            "pvName": "pvc-3a99612e-7fee-46e2-8b56-df8f88b189ec",
            "pvStatus": "Bound",
            "pvcName": "cirros-5-massive-storage-disk-0-gqgzl",
            "workloadsStatus": [
                {
                    "podName": "virt-launcher-cirros-5-massive-storage-mgslr",
                    "podStatus": "Pending",
                    "workloadName": "cirros-5-massive-storage",
                    "workloadType": "VirtualMachineInstance"
                }
            ]
        },
        "lastBackup": "",
        "lastBackupAt": "",
        "lastDegradedAt": "",
        "ownerID": "harv2012",
        "pendingNodeID": "",
        "remountRequestedAt": "",
        "restoreInitiated": false,
        "restoreRequired": false,
        "robustness": "unknown",
        "shareEndpoint": "",
        "shareState": "",
        "state": "detached"
    }
}
```
harv2:~ #

**Comment by w13915984028** [comment]
@yasker @guangbochen   :
Any idea about how to deduce the straightforward error message like "insufficient storage" ?

**Comment by joshimoo** [comment]
On the longhorn side we should prevent the volume creation, if there is no node capable of hosting it during the api volume creation call. We could return the Out of Range error, for this case during the CreateVolume csi call.
This evaluation should be done by the backend api creation call.

ref: https://github.com/container-storage-interface/spec/blob/master/spec.md#createvolume-errors

**Comment by guangbochen** [comment]
@joshimoo will this change be introduced in the LH v1.2.4 or need to wait for the v1.3.0 release, thanks?

**Comment by yasker** [comment]
We shouldn't need to wait for this change. Today Longhorn UI knows if a volume cannot be scheduled according to the conditions of the object. We can just use those conditions to determine.

**Comment by w13915984028** [comment]
> We shouldn't need to wait for this change. Today Longhorn UI knows if a volume cannot be scheduled according to the conditions of the object. We can just use those conditions to determine.

@yasker : 
based on current implementation,
the longhorn pvc has `scheduled ` field, which shows status `False` and reason `ReplicaSchedulingFailure`, with it, we can deduce a general error message like `pvc is not successfully scheduled, VM can not start`, but not `insufficient storage`

is this general error message OK to be shown in VM page, to give the users further information? 

cc @guangbochen 

volumes.longhorn.io pvc-3a99612e-7fee-46e2-8b56-df8f88b189ec

```
            "scheduled": {
                "lastProbeTime": "",
                "lastTransitionTime": "2022-01-12T14:47:10Z",
                "message": "",
                "reason": "ReplicaSchedulingFailure",
                "status": "False",
                "type": "scheduled"
            },
```

**Comment by yasker** [comment]
@joshimoo I remember that condition also indicates we cannot create more replicas with one node. Do you remember how did we tell the two cases apart? Depends on if the volume is attached or not?

**Comment by w13915984028** [comment]
New experiment:
create 2 volumes, but not attaching them to any VM 

@yasker @guangbochen @joshimoo

The first enhancement point with current implementation:
From `PVC ` to `PV ` to `volume (longhorn)`, when it is not schedulable, the first available information can be taken from `volume (longhorn)`, filed `Scheduled`, when its status is "False", we can try to map it into Harvester UI "Volume", to stop the further using of them when creating VM.

Your idea?

And, what is the better way to get this status, harvester controller query longhorn, or longhorn update upstream pvc/pv ?



---
1. they seem to be fine in Harvester->Volumes
![image](https://user-images.githubusercontent.com/31133476/149808389-6b0c49c9-b2fc-40b1-9310-127865a470d7.png)

2. pvc shows OK:
![image](https://user-images.githubusercontent.com/31133476/149808789-7ed1f945-42fa-40cf-809f-a8c3fbc0d387.png)

3. pv shows OK:
![image](https://user-images.githubusercontent.com/31133476/149808695-7627d7b2-d9d8-4095-b6f0-f5d9b62a8cf6.png)

4. volume shows `"Scheduled False"` for the massive-size pv
![image](https://user-images.githubusercontent.com/31133476/149808908-8e3cc384-9de2-4ebb-a34a-0fb5856ce7bc.png)

 
   Scheduled:
      Last Probe Time:
      Last Transition Time:  2022-01-17T16:21:14Z
      Message:
      **Reason:                ReplicaSchedulingFailure
      Status:                False**
      Type:                  scheduled

```
harv2:~ # kubectl describe volume -n longhorn-system pvc-38813022-51ac-4cd0-982d-2f249ccc6e55
Name:         pvc-38813022-51ac-4cd0-982d-2f249ccc6e55
Namespace:    longhorn-system
Labels:       longhornvolume=pvc-38813022-51ac-4cd0-982d-2f249ccc6e55
              recurring-job-group.longhorn.io/default=enabled
Annotations:  <none>
API Version:  longhorn.io/v1beta1
Kind:         Volume
Metadata:
  Creation Timestamp:  2022-01-17T16:21:12Z
  Finalizers:
    longhorn.io
  Generation:  1
  Managed Fields:
    API Version:  longhorn.io/v1beta1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:finalizers:
          .:
          v:"longhorn.io":
        f:labels:
          .:
          f:longhornvolume:
          f:recurring-job-group.longhorn.io/default:
      f:spec:
      f:status:
    Manager:         longhorn-manager
    Operation:       Update
    Time:            2022-01-17T16:21:12Z
  Resource Version:  9629988
  UID:               5dc2535b-436a-4ee3-83d7-52ab3f8eac00
Spec:
  Standby:                    false
  Access Mode:                rwx
  Backing Image:
  Base Image:
  Data Locality:              disabled
  Data Source:
  Disable Frontend:           false
  Disk Selector:              <nil>
  Encrypted:                  false
  Engine Image:               longhornio/longhorn-engine:v1.2.3
  From Backup:
  Frontend:                   blockdev
  Last Attached By:
  Migratable:                 true
  Migration Node ID:
  Node ID:
  Node Selector:              <nil>
  Number Of Replicas:         3
  Replica Auto Balance:       ignored
  Revision Counter Disabled:  false
  Size:                       1072668082176
  Stale Replica Timeout:      30
Status:
  Actual Size:  0
  Clone Status:
    Snapshot:
    Source Volume:
    State:
  Conditions:
    Restore:
      Last Probe Time:
      Last Transition Time:  2022-01-17T16:21:14Z
      Message:
      Reason:
      Status:                False
      Type:                  restore
    Scheduled:
      Last Probe Time:
      Last Transition Time:  2022-01-17T16:21:14Z
      Message:
      Reason:                ReplicaSchedulingFailure
      Status:                False
      Type:                  scheduled
    Toomanysnapshots:
      Last Probe Time:
      Last Transition Time:  2022-01-17T16:21:14Z
      Message:
      Reason:
      Status:                False
      Type:                  toomanysnapshots
  Current Image:             longhornio/longhorn-engine:v1.2.3
  Current Node ID:
  Expansion Required:        false
  Frontend Disabled:         false
  Is Standby:                false
  Kubernetes Status:
    Last PVC Ref At:
    Last Pod Ref At:
    Namespace:           default
    Pv Name:             pvc-38813022-51ac-4cd0-982d-2f249ccc6e55
    Pv Status:           Bound
    Pvc Name:            disk-massive-size
    Workloads Status:    <nil>
  Last Backup:
  Last Backup At:
  Last Degraded At:
  Owner ID:              harv2
  Pending Node ID:
  Remount Requested At:
  Restore Initiated:     false
  Restore Required:      false
  Robustness:            unknown
  Share Endpoint:
  Share State:
  State:                 detached
Events:
  Type    Reason    Age   From                               Message
  ----    ------    ----  ----                               -------
  Normal  Detached  17m   longhorn-volume-controller         volume pvc-38813022-51ac-4cd0-982d-2f249ccc6e55 has been detached
  Normal  Start     17m   longhorn-kubernetes-pv-controller  Persistent Volume pvc-38813022-51ac-4cd0-982d-2f249ccc6e55 started to use/reuse Longhorn volume pvc-38813022-51ac-4cd0-982d-2f249ccc6e55
harv2:~ #
```

**Comment by w13915984028** [comment]
Another related issue
https://github.com/harvester/harvester/issues/1346

It looks, we need to add kind of enhancement to reflect the real status of PVC/PV, and control the usage from UI.
@johnliu55tw @WuJun2016 please also take a look of 1#695

**Comment by w13915984028** [comment]
Place holder: **The second enhancement**:

When Harvester Volume is created via VM creating, in case the Volume is not provisioned by longhorn in certain stage, we need to figure it out and show some useful information in the Web UI.

**Comment by w13915984028** [comment]
A proto-type:

One VM has 2 massive-sized disk, we can try to map the related state from LH volume, to VM annotations (only when there are errors)

Next step, it would need the help of UI team, to display this annotation to proper place.

@yasker @guangbochen @WuJun2016 

![image](https://user-images.githubusercontent.com/31133476/152182455-e2a0c938-1d12-4124-bd08-497e6de523c1.png)

![image](https://user-images.githubusercontent.com/31133476/152182514-a1fbd490-5807-42d8-bd0f-0af333ebdddc.png)

![image](https://user-images.githubusercontent.com/31133476/152182569-f18941c2-0617-40ce-8cd3-9d6ec9e28e29.png)

**Comment by yasker** [comment]
I am fine with using the annotation. Though we need to follow `camel-camel` for the keys, e.g. it should be `vm.harvesterhci.io/volume-status` instead of `volumeStatus`.

**Comment by DaiYuzeng** [comment]
There are two places the annotation could be displayed.

#### Volume tab on VM detail page
<img width="1168" alt="截屏2022-02-15 上午10 03 23" src="https://user-images.githubusercontent.com/15041915/153979273-4e76fde9-21ff-450c-8fd5-75dfa71f2348.png">


#### State column on Volume list
<img width="1139" alt="截屏2022-02-15 上午10 03 34" src="https://user-images.githubusercontent.com/15041915/153979243-b755d9c0-b4e6-4c92-ad38-80e6a7085478.png">

_Mock data_
<img width="996" alt="截屏2022-02-15 上午10 03 07" src="https://user-images.githubusercontent.com/15041915/153979466-f1dc3bd0-630e-442c-8122-c4ea9ae7da53.png">

@w13915984028 Dose it make sense?

**Comment by w13915984028** [comment]
Update:  current backend implementation:
when LH volume is not usable, e.g. because of insufficient storage, the state is updated to PVC and VM via annotations:

`"harvesterhci.io/volume-status"`

when the status is OK, this annotation will be deleted. only error status will be updated.

For VM, it will have info of all its volumes
![image](https://user-images.githubusercontent.com/31133476/154111706-cd1b9640-0edc-4fb8-8d39-597a945a0b2c.png)

For PVC: it will have single info of itself
![image](https://user-images.githubusercontent.com/31133476/154111762-80a4529c-2545-45eb-bbee-3f4ee4bccaf9.png)


At this stage, only `"insufficient storage"`  is abstracted from LH and updated to PVC, VM, in the future, more status could be added.

**Comment by w13915984028** [comment]
The test cases:
to make sure the state is updated with the LH state change, and also with the PVC  attached/detached to/from the VM.

![image](https://user-images.githubusercontent.com/31133476/154113245-1c06ae74-1e44-45c2-8b05-a2bb48c4623b.png)

![image](https://user-images.githubusercontent.com/31133476/154504877-50adc378-ebbc-401d-9fab-f7e05693fa7d.png)

**Comment by w13915984028** [comment]
How to pop up & clear status in a series top down objects:

![image](https://user-images.githubusercontent.com/31133476/154501792-a2b0fcba-4beb-45c9-83ba-e3fa1207ef39.png)

![image](https://user-images.githubusercontent.com/31133476/154502810-5694b2c6-b7a7-4355-9710-ab78ed0e3819.png)


![image](https://user-images.githubusercontent.com/31133476/154502933-9421fb55-8d14-4430-adf7-e21cfcbff18c.png)

![image](https://user-images.githubusercontent.com/31133476/154505433-6d4fe800-b12f-411d-84e5-88a9aa701672.png)

**Comment by WuJun2016** [comment]
The frontend has merged, waiting for the backend code to merge.

**Comment by w13915984028** [comment]
Update: 
Harvester will wait for a solution from Longhorn, then re-adapt. Move back to design now.

@DaiYuzeng @WuJun2016 @guangbochen

**Comment by DaiYuzeng** [comment]
According to https://github.com/longhorn/longhorn/issues/3529#issuecomment-1120670106
I did two tests:

1. Create a volume exceeded maximum storage.
2. Create a volume and then resizie it to exceed maximum storage.

#### situation of test 1
There is no related pv and longhorn volume

#### situation of test 2
There are related pv and longhorn volume, but there is no error message at the annotation ` longhorn.io/volume-scheduling-error` of pv.


@guangbochen @futuretea @WuJun2016 ^^

**Comment by guangbochen** [comment]
Tentatively move this to v1.1.1 for now and need to wait for the upstream fix from the LH team, thanks.

**Comment by TachunLin** [comment]
Verified **fixed** on `master-b0d883ce-head` (11/22). Close this issue.

### Result
#### Case 1 (PASS)
1. When we create volume exceed the max acceptable size, will block and prompt message
`Exceed maximum size `999999999` Gi!`
  ![image](https://user-images.githubusercontent.com/29251855/203491945-16666082-f17d-406e-97df-e6f75a5fa377.png)

#### Case 2 (PASS)
1. When we create a volume did not exceed but larger than the total disk capability (plus overcommit), 
2. We can create the volume but the status of the volume is `NotReady`
3. The error message should be insufficient storage
  ![image](https://user-images.githubusercontent.com/29251855/203516617-f4a206ad-8cd3-4ff4-a579-7dcb12bcd644.png)
4. The link on the top of the page will be visible
  ![image](https://user-images.githubusercontent.com/29251855/203516769-8540e058-63fe-4649-9257-577fc0b15de9.png)
5. Click the link to open a new tab with embedded longhorn UI 
  ![image](https://user-images.githubusercontent.com/29251855/203516850-78053df9-35aa-46d6-ae70-7df74ec04ede.png)


#### Case 3 single node (PASS)
When we create a vm have os image volume with `replica scheduling failed` and another volume with `insufficient disk`
1.  The status of the volume should be `Degraded` and `Not ready`
2.  The error message of the volume should be `replica scheduling failed` and `insufficient disk`
  ![image](https://user-images.githubusercontent.com/29251855/203511434-190212bd-fdd6-40cf-b79a-225bb859160b.png)
3. The link in the volume section will be visible


#### Case 4 multiple nodes (PASS)
Same result as single node but in multi nodes cluster
  ![image](https://user-images.githubusercontent.com/29251855/203514630-c25edb9c-5bb6-4072-b200-8aa57e68da81.png)


### Test Information
* Test Environment: 3 nodes harvester on bare machines and 1 nodes local kvm machine
* Harvester version:`master-b0d883ce-head` (11/22)

### Verify Steps
### Case 1

1. Go to create a volume with 9999999999 Gi storage size in both harvester-longhorn and longhorn storage class
2. The volume should not be created successfully.
3. The error message should be Exceed maximum size `999999999` Gi!

### Case 2 

1. Go to create a volume with 99999999 Gi storage size in both harvester-longhorn and longhorn storage class
2. Back to volume list
3. The status of the volume should be NotReady
4. The error message should be insufficient storage
5. Go to preference page and enable DEV mode
6. Go to volume detail page
7. The link on the top of the page will be visible
8. Click the link to open a new tab with embedded longhorn UI 

### Case 3 (Single node)
1. Go to create a vm 
2. Give os image volume 30Gi and add another volume 1600Gi of harvester-longhorn storage class
![image](https://user-images.githubusercontent.com/29251855/203510571-301e1ce9-a8f8-45bf-8320-44e0d267a02e.png)
3. Go to vm detail or vm edit page
4. Switch to volume tab
5. The status of the volume should be `Degraded` and `Not ready`
6. The error message of the volume should be `replica scheduling failed` and `insufficient disk`
![image](https://user-images.githubusercontent.com/29251855/203511434-190212bd-fdd6-40cf-b79a-225bb859160b.png)
7. The link in the volume section will be visible

### Case 4 (multi nodes)

1. Disable `Node scheduling` on node 3 in Longhorn UI
  ![image](https://user-images.githubusercontent.com/29251855/203519240-a4763d22-5e6a-44bb-9b33-14aef0f80a1e.png)
  ![image](https://user-images.githubusercontent.com/29251855/203514056-5af3d7db-b9e1-4603-90b0-b366dc300c64.png)

2. Give os image volume 30Gi and add another volume 1600Gi of harvester-longhorn storage class
  ![image](https://user-images.githubusercontent.com/29251855/203514274-79240792-aaed-4c63-a407-4a7a5b78930c.png)

3. OS image disk display `degraded` while another disk display `Not ready` with corresponding error
  ![image](https://user-images.githubusercontent.com/29251855/203514630-c25edb9c-5bb6-4072-b200-8aa57e68da81.png)
